Use target and metadata from context

gouthamve opened the pull request:
**Description:** This uses the target and metadata info passed in through context instead of looking them up via `instance` and `job` labels.

**Link to tracking Issue:** This fixes #5757 and #5663

**Testing:** I've done a `Prometheus --> OTel Collector --> Another Prometheus` with `honor_labels: true` and Prometheus federation. It used to fail before but now it works!

**TODO**:
- [x] Get https://github.com/prometheus/prometheus/pull/10473 merged
- [x] Fix tests / add new ones

gouthamve committed:
Use target and metadata from context

This fixes #5757 and #5663

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

gouthamve commented:
I'll resolve the conflicts once upstream PR is merged.

dashpole commented on receiver/prometheusreceiver/internal/otlp_transaction_test.go:
```diff
@@ -14,125 +14,125 @@
-	"context"
-	"testing"
-	"time"
+// import (
```
I assume you are still in the process of fixing these tests?

dashpole commented:
Looks good once the prerequisites are complete

gouthamve committed:
Add tests for relabeling working

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

gouthamve committed:
Use Prometheus main branch

https://github.com/prometheus/prometheus/pull/10473 has been merged

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

pmm-sumo commented on receiver/prometheusreceiver/internal/otlp_transaction_test.go:
```diff
@@ -14,125 +14,125 @@
-			Value: "localhost:8080",
-		},
-	)
+// func Test_transaction_pdata(t *testing.T) {
```
Why this (and other) tests are commented?

gouthamve replied:
hi, sorry, missed this somehow. Fixed it now!

pmm-sumo commented on receiver/prometheusreceiver/metrics_receiver.go:
```diff
@@ -90,7 +90,6 @@ func (r *pReceiver) Start(_ context.Context, host component.Host) error {
 		r.cfg.PrometheusConfig.GlobalConfig.ExternalLabels,
 	)
 	r.scrapeManager = scrape.NewManager(&scrape.Options{}, logger, r.ocaStore)
-	r.ocaStore.SetScrapeManager(r.scrapeManager)
```
I think ScrapeManager capability has been removed from prometheusreceiver. Basing on the comments in other section, it's needed for "OCAStore". I'm not sure if the latter is still relevant but wanted to make sure we're not removing some part of functionality

Aneurysm9 replied:
Strike that, reverse it.  This PR removes the need for the scrape manager in the OCAStore, but the receiver still needs it as it is what handles scraping targets emitted by the discovery manager.

gouthamve replied:
I'm not sure I follow but as Anthony said, its not required for `OCAStore` anymore. But it still exists in `r.scrapeManager`. I am resolving this comment as I think it has been addressed, but please let me know if I am missing something.

pmm-sumo commented:
I think the rebase resulted in a bunch of `go.mod`/`go.sum` modified in components unrelated to changes

Aneurysm9 commented on receiver/prometheusreceiver/internal/instance_relabel_end_to_end_test.go:
```diff
@@ -0,0 +1,224 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver"
+)
+
+func TestJobInstanceRelabelEndToEnd(t *testing.T) {
```
This should probably happen in the `prometheusreceiver` package where we already have a framework for e2e tests and shape testing of the resulting pdata metrics.

gouthamve replied:
Hrm, so, I tried moving it there, but the framework there is about verifying the "scrape" output and I cannot figure out how to verify the target and job labels that are added. Could you give me a pointer?

Aneurysm9 replied:
You'd want to build a [test like this one](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b00283b724e83bb91526108cdc3c43520c167961/receiver/prometheusreceiver/metrics_receiver_labels_test.go#L31), probably in that same file since it's already testing labeling handling.  The [validation function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b00283b724e83bb91526108cdc3c43520c167961/receiver/prometheusreceiver/metrics_receiver_labels_test.go#L47) can prepare a [set of expected resource attributes](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b00283b724e83bb91526108cdc3c43520c167961/receiver/prometheusreceiver/metrics_receiver_labels_test.go#L47), which in this case would include the job and instance ID values mapped to the appropriate resource attribute identifiers.  `doCompare()` will [take care of asserting](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/519ea7ff61cb09d2be2d40b070fad753a1603a9b/receiver/prometheusreceiver/metrics_receiver_helper_test.go#L326-L332) that all of the expected resource attributes are present.

gouthamve replied:
Ah ack, working on it now. I didn't realise the resource attributes part.

gouthamve committed:
Add back the tests

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

gouthamve committed:
Fix flaky test

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

gouthamve committed:
Add Changelog entry

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

gouthamve committed:
Add relabel test with the e2e framework

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

gouthamve commented:
@dashpole @Aneurysm9 @pmm-sumo Thanks for the initial review, I think I have addressed all the comments now. Could you take another look when you have some time?

Aneurysm9 commented on receiver/prometheusreceiver/metrics_receiver_labels_test.go:
```diff
@@ -619,3 +620,70 @@ func TestHonorLabelsTrueConfig(t *testing.T) {
+					Replacement: "relabeled-instance",
+				},
+				{
+					// this config should replace the instance label with 'relabeled-instance'
```
```suggestion
					// this config should replace the job label with 'not-target1'
```

gouthamve committed:
Update receiver/prometheusreceiver/metrics_receiver_labels_test.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

jpkrohling committed:
Move changelog entry to unreleased

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling commented:
PR updated to account for the v0.48.0 changes

gouthamve committed:
Make lint pass

Needed to run make gotidy; make golint

strings.Title is deprecated

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

jpkrohling commented:
Nice, github now allows us to restart only the failed jobs. Once we have a green build, I'll merge this.

shacharSirotkin commented:
Hi,
I know it's a pretty old PR but I would be happy if you can upload a working config for Prometheus federation. 
I'm struggling with it for a long time. 
Thanks 🙏

mx-psi commented:
> Hi, I know it's a pretty old PR but I would be happy if you can upload a working config for Prometheus federation. I'm struggling with it for a long time. Thanks 🙏

Please open a new issue for this, thanks!
